packages/eslint-plugin/src/rules/no-unnecessary-generics.ts

```diff
@@ -0,0 +1,315 @@
+            const res = getSoleUse(
+              sig,
+              assertDefined(checker.getSymbolAtLocation(tp.name)),
+              checker,
+            );
+            switch (res.type) {
+              case 'ok':
+                break;
+              case 'sole':
+                context.report({
+                  node: parserServices.tsNodeToESTreeNodeMap.get(res.soleUse),
+                  messageId: 'failureString',
+                  data: {
+                    typeParameter,
+                  },
+                });
+                break;
+              case 'never':
+                context.report({
+                  node: parserServices.tsNodeToESTreeNodeMap.get(tp),
+                  messageId: 'failureString',
+                  data: {
+                    typeParameter,
+                  },
+                });
+                break;
+              default:
+                assertNever(res);
+            }
+          }
+        }
+      },
+    };
+  },
+});
+
+type Result =
+  | { type: 'ok' | 'never' }
+  | { type: 'sole'; soleUse: ts.Identifier };
+function getSoleUse(
+  sig: ts.SignatureDeclaration,
+  typeParameterSymbol: ts.Symbol,
+  checker: ts.TypeChecker,
+): Result {
+  const exit = {};
+  let soleUse: ts.Identifier | undefined;
+  const seenTypes = new Set<ts.Type>();
+
+  try {
+    if (sig.typeParameters) {
+      for (const tp of sig.typeParameters) {
+        if (tp.constraint) {
+          recur(tp.constraint);
+        }
+      }
+    }
+    for (const param of sig.parameters) {
+      if (param.type) {
+        recur(param.type);
+      }
+    }
+    if (sig.type) {
+      recur(sig.type);
+    } else if (sig.kind !== ts.SyntaxKind.Constructor && 'body' in sig) {
+      // if body is missing, there is no point inferring return type.
+      const sigType = checker.getSignatureFromDeclaration(sig);
+      if (!sigType) {
+        return { type: 'ok' };
+      }
+
+      const returnType = checker.getReturnTypeOfSignature(sigType);
+      try {
+        forEachType(
+          returnType,
+          type => {
+            if (type.getSymbol() === typeParameterSymbol) {
+              throw exit;
+            }
+          },
+          checker,
+        );
+      } catch (error) {
+        if (error === exit) {
+          return { type: 'ok' };
+        } else {
+          throw error;
+        }
+      }
+    }
+  } catch (err) {
+    if (err === exit) {
+      return { type: 'ok' };
+    }
+    throw err;
+  }
+
+  return soleUse ? { type: 'sole', soleUse } : { type: 'never' };
+
+  function recur(node: ts.TypeNode): void {
+    if (ts.isIdentifier(node)) {
+      if (checker.getSymbolAtLocation(node) === typeParameterSymbol) {
+        if (soleUse === undefined) {
+          soleUse = node;
+        } else {
+          throw exit;
+        }
+      }
+    } else {
+      node.forEachChild(recur as (node: ts.Node) => void);
+    }
+  }
+
+  function forEachType(
```

Comment:
>   ● no-unnecessary-generics › valid › 
> ```ts
> function f1<T extends { new (...args: any[]): any }>(a: T) {
>   return class Foo extends a {
>     f = 1;
>   };
> }
> ```    
> 
>     RangeError: Maximum call stack size exceeded
>     Occurred while linting typescript-eslint/packages/eslint-plugin/tests/fixtures/file.ts:2
> 
>       250 |         });
>       251 |
>     > 252 |         forEachType(signature.getReturnType(), cb, checker);
>           |                               ^
>       253 |
>       254 |         // const predicate = checker.getTypePredicateOfSignature();
>       255 |         const decl = signature?.getDeclaration();

It seems checking if a type is used in another type is extremely difficult. Even an innocent constructor signature may kill the linter.

Reply:
Is there another way to get this info?
There must be otherwise TS would always die for that sort of code.